apps/sms/views/shared/js/startup.js

```diff
@@ -89,14 +89,15 @@ var Startup = exports.Startup = {
 
       // Init UI Managers
       TimeHeaders.init();
-      ConversationView.init();
-      MessagingClient.init(App.instanceId);
       MozMobileConnectionsClient.init(App.instanceId);
-      Information.initDefaultViews();
+      MessagingClient.init(App.instanceId).then(() => {
```

Comment:
I'm wondering why we should wait until ```MessagingService``` confirms that event listeners are registered? Can't we just call ```MessagingService.register()``` (maybe we can choose more generic method name like 'register' instead of 'registerEvents') and don't wait for the return result?

Your approach looks good enough for me, the only alternative option (it's likely worse :), but I put it on the table anyway) I had in mind is to invert dependency here: once iframe/alwaysLowered-window/shim is created it establishes connection with MessagingService by itself and transmits events via this dedicated _pipe_ (including appInstanceId to indicate target for the event). MessagingService _optionally_ can have additional dedicated internal service to handle this _event pipe_. But as I said in the beginning it looks awkward and complex :p